### ISSUE TITLE
Assess facet request thresholds

### ISSUE THREAD

**Issue by brent-hartwig** [project-lux/lux-marklogic#179]
**Problem Description:**
Thresholds are in place to protect the system.  When the facet response body changed to Activity Streams, the 'facet maximum product' threshold probably should have been reduced (or changed to different logic).  Now that we paginated facet requests, we have a second reason to review this threshold.

Scope is to review **all** facet-related thresholds and adjust to align with the current implementation, striking a balance between conserving system resources, maintaining system stability, and providing end-uses access to all data.

Out of scope: thresholds for non-paginated facet requests.  Our current direction is to always paginate but should that change, the scope of another ticket can handle those threshold changes.

**Expected Behavior/Solution:**
Review the following thresholds and determine if any changes should be made.

1. Non-semantic facet thresholds:
    * When the search result estimate multiplied by the associated index's value count exceeds a threshold set by the `facetMaximumProduct` property, the facet request is rejected.
    * 20 seconds
2. Semantic facet thresholds:
    * Requests are limited to 100 values (hard-coded).
    * 59 seconds

**Requirements:**
See above.

**Needed for promotion:**
If an item on the list is not needed, it should be crossed off but not removed.

~- [ ] Wireframe/Mockup - Mike~
- [ ] Committee discussions - Sarah
- [ ] Feasibility/Team discussion - Sarah
- [ ] Backend requirements - TBD
~- [ ] Frontend requirements - no impact to the frontend is expected~
- [ ] Are new regression tests required for QA - Amy
- [ ] Compels a performance test?
- [ ] Questions
- List of questions for discussions. Answers should be documented within the issue.

**UAT/LUX Examples:**
Developer to specify all impacted types of facet requests.

**Dependencies/Blocks:**
This ticket should follow #160 and #161.  ~If those are held up too long, we should consider updating the 'facet product maximum' threshold sooner as the Activity Streams change (https://git.yale.edu/lux-its/marklogic/issues/1164) is already in production and can increase the response time by 3.1x and the response body size by 9.4x.~  --both have been implemented.

**Related Github Issues:**

* #56 would like us to reject fewer facet requests.
* Paginated facet request's performance test: #162 --facet pagination passed its performance test.

**Related links:**
None.

**Wireframe/Mockup:**
Not needed.

**Comment by brent-hartwig** [comment]
cc: @roamye, @clarkepeterf

**Comment by brent-hartwig** [comment]
@clarkepeterf, I noticed the backend took over seven seconds to before rejecting a couple production facet requests due to exceeding the maximum product threshold.  Unless something else leading up to the check is consuming this time, I have to believe it is the check's call to `cts.estimate`, which would make it too expensive of a check.  Time spent in `cts.estimate` takes from the 20 second timeout.  I'm more inclined to remove this check and only provide an alternative if we need to.  If we choose to remove it:

1. We can also remove the field count data constants.
2. We should probably do so as the scope of a separate ticket, leaving this one open (unless we want to take on the rest of this ticket's scope at the same time).

```
2024-07-30 18:39:34.831 Info: [Event:id=LuxFacets] Rejected request to calculate the workCreationAgentId facet as 1,352,187 search results by 4,992,982 field values exceeds the 5,000,000,000,000 threshold.
2024-07-30 18:39:34.831 Info: [Event:id=LuxFacets] Failed to calculate the following facet after 7325 milliseconds: workCreationAgentId (page: 1; pageLength: 20; filterResults: n/a)
2024-07-30 18:39:34.832 Notice: JS-JAVASCRIPT: throw e; -- Error running JavaScript request: Error: Threshold to calculate the workCreationAgentId facet exceeded.
2024-07-30 18:39:34.832 Notice:+in ./mlErrorsLib.mjs, at 9:4, in BadRequestError() [javascript]
2024-07-30 18:39:34.832 Notice:+in ../../lib/facetsLib.mjs, at 265:10, in _getNonSemanticFacet() [javascript]
2024-07-30 18:39:34.832 Notice:+in ../../lib/facetsLib.mjs, at 86:37, in getFacet() [javascript]
2024-07-30 18:39:34.832 Notice:+in /ds/lux/facets.mjs, at 13:9 [javascript]
2024-07-30 18:39:34.832 Notice:+in ../../lib/requestHandleLib.mjs, at 10:9, in handleRequest() [javascript]
2024-07-30 18:39:34.832 Notice:+in /ds/lux/facets.mjs, at 12:0 [javascript]
2024-07-30 18:39:34.832 Notice:+in /ds/lux/facets.mjs [javascript]
2024-07-30 18:39:34.835 Warning: {"errorResponse":{"statusCode":400,"status":"Bad Request","messageCode":"BadRequestError","message":"Threshold to calculate the workCreationAgentId facet exceeded."}}
```

```
2024-07-30 18:39:34.995 Info: [Event:id=LuxFacets] Rejected request to calculate the workAboutConceptId facet as 1,352,187 search results by 4,310,211 field values exceeds the 5,000,000,000,000 threshold.
2024-07-30 18:39:34.996 Info: [Event:id=LuxFacets] Failed to calculate the following facet after 7483 milliseconds: workAboutConceptId (page: 1; pageLength: 20; filterResults: n/a)
2024-07-30 18:39:34.997 Notice: JS-JAVASCRIPT: throw e; -- Error running JavaScript request: Error: Threshold to calculate the workAboutConceptId facet exceeded.
2024-07-30 18:39:34.997 Notice:+in ./mlErrorsLib.mjs, at 9:4, in BadRequestError() [javascript]
2024-07-30 18:39:34.997 Notice:+in ../../lib/facetsLib.mjs, at 265:10, in _getNonSemanticFacet() [javascript]
2024-07-30 18:39:34.997 Notice:+in ../../lib/facetsLib.mjs, at 86:37, in getFacet() [javascript]
2024-07-30 18:39:34.997 Notice:+in /ds/lux/facets.mjs, at 13:9 [javascript]
2024-07-30 18:39:34.997 Notice:+in ../../lib/requestHandleLib.mjs, at 10:9, in handleRequest() [javascript]
2024-07-30 18:39:34.997 Notice:+in /ds/lux/facets.mjs, at 12:0 [javascript]
2024-07-30 18:39:34.997 Notice:+in /ds/lux/facets.mjs [javascript]
2024-07-30 18:39:35.033 Warning: {"errorResponse":{"statusCode":400,"status":"Bad Request","messageCode":"BadRequestError","message":"Threshold to calculate the workAboutConceptId facet exceeded."}}
```

cc: @roamye, @jffcamp

**Comment by brent-hartwig** [comment]
Pinging Mr. @clarkepeterf ☝️

**Comment by clarkepeterf** [comment]
I would like to investigate if the issue is with cts.estimate

**Comment by clarkepeterf** [comment]
@brent-hartwig do you know what search criteria was used for these requests?

**Comment by clarkepeterf** [comment]
After more consideration I think it makes sense to just remove this threshold

**Comment by brent-hartwig** [comment]
@clarkepeterf, I started looking for the search criteria but yeah, I'm good with it being removed 💯

**Comment by clarkepeterf** [comment]
To be removed via https://github.com/project-lux/lux-marklogic/issues/290

**Comment by roamye** [comment]
Can this issue be proposed closed as we have assessed and determined the request thresholds should be removed? Or is there something else needed here? 

@brent-hartwig @clarkepeterf

**Comment by clarkepeterf** [comment]
@roamye Ok to close

**Comment by brent-hartwig** [comment]
@clarkepeterf and @roamye, please note the scope of this ticket was for all facet-related thresholds, not just the one we agreed to remove.  If we're good with the other three and do not need to consider a new one, #proposeClose.

**Comment by clarkepeterf** [comment]
@brent-hartwig good point.

#290 will remove the non-semantic threshold.

There is a semantic threshold which I'm not sure if we should remove.

Is there a third threshold?

cc @roamye

**Comment by brent-hartwig** [comment]
The four thresholds are listed in the Expected Behavior/Solution section.

**Comment by clarkepeterf** [comment]
@brent-hartwig I see, I was just looking at the thresholds based on number of values. You also mean the timeouts?

So, if we remove those, they would just be the default timeout for the app server? (currently 20 seconds)

**Comment by brent-hartwig** [comment]
Yes, timeouts too.  For facets, we only override the default timeout for semantic facets.  Non-semantic facets are subject to the default, which is presently 20 seconds.  I have conflicting thoughts on the timeouts, including:

1. The original intent was to conserve system resources by not allowing any one user's actions to consume resources for what we consider an excessive amount of time.
2. We're not experiencing concurrent demand issues in prod.
3. We're now paginating non-semantic facet requests --this should reduce the risk of removing (and not replacing) the current maximum product threshold.
4. The default is generous for the vast majority of requests.
5. Users are only going to wait around so long anyway.

I don't think we should reduce the semantic facet timeout given semantic facet requests can still timeout after 59 seconds when the associated search returns many search results.  I don't believe our current two semantic facets hit the 100 maximum value threshold, and am inclined to keep it for the time being.

In summary, we may be just fine with the other three thresholds but probably good to document some thoughts on them before closing this ticket.

**Comment by clarkepeterf** [comment]
I am good with keeping the semantic facet threshold and timeout. non-semantic facets have default timeout and non-semantic threshold is to be removed by #290. Therefore, I am good with #propose-close on this ticket.

**Comment by roamye** [comment]
Approved by UAT to close
